userapi/src/main/java/fk/prof/userapi/Deserializer.java

```diff
@@ -0,0 +1,65 @@
+package fk.prof.userapi;
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.zip.Adler32;
+
+/**
+ * @author gaurav.ashok
+ */
+public abstract class Deserializer<T> {
+    abstract public T deserialize(InputStream in);
+
+    public static int readInt32(InputStream in) throws IOException {
+        int ch1 = in.read();
+        int ch2 = in.read();
+        int ch3 = in.read();
+        int ch4 = in.read();
+        if ((ch1 | ch2 | ch3 | ch4) < 0)
+            throw new EOFException();
+        return ((ch1 << 24) + (ch2 << 16) + (ch3 << 8) + (ch4 << 0));
+    }
+
+    /**
+     * Read a series of bytes that was serialized in the format of [size][message][checksum].
+     * @param in
+     * @return
+     * @throws IOException
+     */
+    public static byte[] readFenced(InputStream in) throws IOException {
+        // read size
+        int size = readInt32(in);
+        // read size bytes
+        byte[] bytes = new byte[size];
+        readFully(in, bytes);
```

Comment:
You many not have read enough to fill the byte-array up. Its better to fail with a logical exception saying "we couldn't read enough" vs. mysterious checksum-verification-failure. 

So check if actual-read (return value) equals bytes.length.